AbstractCliInternetAddressResolver attempts to execute commands that may not exist

##### Issue Overview

I'm using Arquillian Cube in my testsuite together with the `@RequiresDocker` annotation. All seems to be well on *nix. 

On Windows I hit a problem if no Docker client is available. `AbstractCliInternetAddressResolver.getIp()` is trying to execute `boot2docker`.

Should it not be checking that the target executable is available before attempting to execute it? It'd make more sense from a 'RequiresDocker' standpoint (I.e the target client executable should be present to fulfil the requirement check).

##### Expected Behaviour

Test is skipped because no Docker client is available (or runs as normal if one is found).

##### Current Behaviour

Exception. Stack trace:

```
Caused by: org.arquillian.spacelift.execution.ExecutionException: Execution of a task failed. Execution of a task failed. Unable to execute task SpawnProcessTask
	at org.arquillian.spacelift.execution.ExecutionException.prependMessage(ExecutionException.java:70)
	at org.arquillian.spacelift.execution.impl.FutureBasedExecution.unwrapException(FutureBasedExecution.java:173)
	at org.arquillian.spacelift.execution.impl.FutureBasedExecution.await(FutureBasedExecution.java:101)
	at org.arquillian.spacelift.task.os.ProcessBasedExecution.await(ProcessBasedExecution.java:149)
	at org.arquillian.cube.docker.impl.util.CommandLineExecutor.execCommandAsArray(CommandLineExecutor.java:22)
	at org.arquillian.cube.docker.impl.util.CommandLineExecutor.execCommand(CommandLineExecutor.java:14)
	at org.arquillian.cube.docker.impl.util.AbstractCliInternetAddressResolver.getIp(AbstractCliInternetAddressResolver.java:29)
	at org.arquillian.cube.docker.impl.util.AbstractCliInternetAddressResolver.ip(AbstractCliInternetAddressResolver.java:23)
	at org.arquillian.cube.docker.impl.client.CubeDockerConfigurationResolver.resolveBoot2Docker(CubeDockerConfigurationResolver.java:283)
	at org.arquillian.cube.docker.impl.client.CubeDockerConfigurationResolver.resolveServerIp(CubeDockerConfigurationResolver.java:190)
	at org.arquillian.cube.docker.impl.client.CubeDockerConfigurationResolver.resolve(CubeDockerConfigurationResolver.java:65)
	at org.arquillian.cube.docker.impl.client.CubeDockerConfigurator.configure(CubeDockerConfigurator.java:87)
	at org.arquillian.cube.docker.impl.client.CubeDockerConfigurator.configure(CubeDockerConfigurator.java:77)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:94)
	at org.jboss.arquillian.core.impl.EventContextImpl.invokeObservers(EventContextImpl.java:99)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:81)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:145)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:116)
	at org.jboss.arquillian.core.impl.ManagerImpl.bindAndFire(ManagerImpl.java:265)
	at org.jboss.arquillian.core.impl.InstanceImpl.set(InstanceImpl.java:74)
	at org.arquillian.cube.impl.client.CubeConfigurator.configure(CubeConfigurator.java:24)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:94)
	at org.jboss.arquillian.core.impl.EventContextImpl.invokeObservers(EventContextImpl.java:99)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:81)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:145)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:116)
	at org.jboss.arquillian.core.impl.ManagerImpl.bindAndFire(ManagerImpl.java:265)
	at org.jboss.arquillian.core.impl.InstanceImpl.set(InstanceImpl.java:74)
	at org.jboss.arquillian.config.impl.extension.ConfigurationRegistrar.loadConfiguration(ConfigurationRegistrar.java:73)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:94)
	at org.jboss.arquillian.core.impl.EventContextImpl.invokeObservers(EventContextImpl.java:99)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:81)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:145)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:116)
	at org.jboss.arquillian.core.impl.ManagerImpl.start(ManagerImpl.java:290)
	at org.jboss.arquillian.test.impl.EventTestRunnerAdaptor.<init>(EventTestRunnerAdaptor.java:63)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.jboss.arquillian.test.spi.SecurityActions.newInstance(SecurityActions.java:161)
	at org.jboss.arquillian.test.spi.SecurityActions.newInstance(SecurityActions.java:102)
	at org.jboss.arquillian.test.spi.TestRunnerAdaptorBuilder.build(TestRunnerAdaptorBuilder.java:52)
	at org.jboss.arquillian.junit.Arquillian.run(Arquillian.java:113)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:283)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:173)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:153)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:128)
	at org.apache.maven.surefire.booter.ForkedBooter.invokeProviderInSameClassLoader(ForkedBooter.java:203)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:155)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:103)
Caused by: java.io.IOException: Cannot run program "boot2docker": CreateProcess error=2, The system cannot find the file specified
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
	at org.arquillian.spacelift.task.os.SpawnProcessTask.process(SpawnProcessTask.java:107)
	at org.arquillian.spacelift.task.os.SpawnProcessTask.process(SpawnProcessTask.java:34)
	at org.arquillian.spacelift.task.Task.run(Task.java:98)
	at org.arquillian.spacelift.task.Task$1.call(Task.java:70)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.io.IOException: CreateProcess error=2, The system cannot find the file specified
	at java.lang.ProcessImpl.create(Native Method)
	at java.lang.ProcessImpl.<init>(ProcessImpl.java:386)
	at java.lang.ProcessImpl.start(ProcessImpl.java:137)
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
	at org.arquillian.spacelift.task.os.SpawnProcessTask.process(SpawnProcessTask.java:107)
	at org.arquillian.spacelift.task.os.SpawnProcessTask.process(SpawnProcessTask.java:34)
	at org.arquillian.spacelift.task.Task.run(Task.java:98)
	at org.arquillian.spacelift.task.Task$1.call(Task.java:70)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

##### Steps To Reproduce
1. Annotate a test case with `@RequiresDocker` then run on Windows without a Docker client installed.

--- a/docker/docker/src/main/java/org/arquillian/cube/docker/impl/requirement/DockerMachineRequirement.java
+++ b/docker/docker/src/main/java/org/arquillian/cube/docker/impl/requirement/DockerMachineRequirement.java
@@ -8,7 +8,15 @@ import org.arquillian.spacelift.execution.ExecutionException;
 
 public class DockerMachineRequirement implements Requirement<RequiresDockerMachine> {
 
-    private final CommandLineExecutor commandLineExecutor = new CommandLineExecutor();
+    private final CommandLineExecutor commandLineExecutor;
+
+    public DockerMachineRequirement() {
+        commandLineExecutor = new CommandLineExecutor();
+    }
+
+    public DockerMachineRequirement(CommandLineExecutor commandLineExecutor) {
+        this.commandLineExecutor = commandLineExecutor;
+    }
 
     @Override
     public void check(RequiresDockerMachine context) throws UnsatisfiedRequirementException {

--- a/docker/docker/src/main/java/org/arquillian/cube/docker/impl/requirement/DockerRequirement.java
+++ b/docker/docker/src/main/java/org/arquillian/cube/docker/impl/requirement/DockerRequirement.java
@@ -1,11 +1,13 @@
 package org.arquillian.cube.docker.impl.requirement;
 
+import java.util.HashMap;
+import java.util.Map;
+
 import com.github.dockerjava.api.DockerClient;
 import com.github.dockerjava.api.model.Version;
 import com.github.dockerjava.core.DockerClientBuilder;
 import com.google.common.base.Strings;
-import java.util.HashMap;
-import java.util.Map;
+
 import org.arquillian.cube.docker.impl.client.CubeDockerConfiguration;
 import org.arquillian.cube.docker.impl.client.CubeDockerConfigurationResolver;
 import org.arquillian.cube.docker.impl.util.Boot2Docker;
@@ -15,15 +17,26 @@ import org.arquillian.cube.docker.impl.util.OperatingSystemResolver;
 import org.arquillian.cube.docker.impl.util.Top;
 import org.arquillian.cube.spi.requirement.Requirement;
 import org.arquillian.cube.spi.requirement.UnsatisfiedRequirementException;
+import org.arquillian.spacelift.execution.ExecutionException;
 
 public class DockerRequirement implements Requirement<RequiresDocker> {
 
-    private final CommandLineExecutor commandLineExecutor = new CommandLineExecutor();
-    private final CubeDockerConfigurationResolver resolver = new CubeDockerConfigurationResolver(new Top(),
-        new DockerMachine(commandLineExecutor),
-        new Boot2Docker(commandLineExecutor),
-        new OperatingSystemResolver().currentOperatingSystem().getFamily()
-    );
+    private final CommandLineExecutor commandLineExecutor;
+    private final CubeDockerConfigurationResolver resolver;
+
+    public DockerRequirement() {
+        this.commandLineExecutor = new CommandLineExecutor();
+        this.resolver = new CubeDockerConfigurationResolver(new Top(),
+            new DockerMachine(commandLineExecutor),
+            new Boot2Docker(commandLineExecutor),
+            new OperatingSystemResolver().currentOperatingSystem().getFamily()
+        );
+    }
+
+    public DockerRequirement(CommandLineExecutor commandLineExecutor, CubeDockerConfigurationResolver resolver) {
+        this.commandLineExecutor = commandLineExecutor;
+        this.resolver = resolver;
+    }
 
     /**
      * @param serverUrl
@@ -52,12 +65,18 @@ public class DockerRequirement implements Requirement<RequiresDocker> {
 
     @Override
     public void check(RequiresDocker context) throws UnsatisfiedRequirementException {
-        Map<String, String> config = resolver.resolve(new HashMap<String, String>());
-        String serverUrl = config.get(CubeDockerConfiguration.DOCKER_URI);
-        if (Strings.isNullOrEmpty(serverUrl)) {
-            throw new UnsatisfiedRequirementException("Could not resolve the docker server url.");
-        } else if (!isDockerRunning(serverUrl)) {
-            throw new UnsatisfiedRequirementException("No server is running on url:[" + serverUrl + "].");
+        try {
+            Map<String, String> config = resolver.resolve(new HashMap<String, String>());
+            String serverUrl = config.get(CubeDockerConfiguration.DOCKER_URI);
+            if (Strings.isNullOrEmpty(serverUrl)) {
+                throw new UnsatisfiedRequirementException("Could not resolve the docker server url.");
+            } else if (!isDockerRunning(serverUrl)) {
+                throw new UnsatisfiedRequirementException("No server is running on url:[" + serverUrl + "].");
+            }
+        } catch (ExecutionException e) {
+            throw new UnsatisfiedRequirementException("Cannot execute docker command.");
         }
     }
+
+
 }

--- a/docker/docker/src/test/java/org/arquillian/cube/docker/impl/requirement/DockerMachineRequirementTest.java
+++ b/docker/docker/src/test/java/org/arquillian/cube/docker/impl/requirement/DockerMachineRequirementTest.java
@@ -1,0 +1,76 @@
+package org.arquillian.cube.docker.impl.requirement;
+
+import java.lang.annotation.Annotation;
+import java.util.Arrays;
+import java.util.Collections;
+
+import org.arquillian.cube.docker.impl.util.CommandLineExecutor;
+import org.arquillian.cube.spi.requirement.UnsatisfiedRequirementException;
+import org.arquillian.spacelift.execution.ExecutionException;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+import static org.mockito.Matchers.anyVararg;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class DockerMachineRequirementTest {
+
+    @Mock
+    CommandLineExecutor commandLineExecutor;
+
+    @Test(expected = UnsatisfiedRequirementException.class)
+    public void estDockerMachineRequirementCheckWhenExecutionExceptionThrown() throws UnsatisfiedRequirementException {
+        when(commandLineExecutor.execCommandAsArray(anyVararg())).thenThrow(ExecutionException.class);
+
+        DockerMachineRequirement dockerMachineRequirement = new DockerMachineRequirement(commandLineExecutor);
+        dockerMachineRequirement.check(createContext("testing"));
+    }
+
+    @Test(expected = UnsatisfiedRequirementException.class)
+    public void testDockerMachineRequirementCheckNoMatchingNameFound() throws Exception {
+        when(commandLineExecutor.execCommandAsArray(anyVararg())).thenReturn(Collections.emptyList());
+
+        DockerMachineRequirement dockerMachineRequirement = new DockerMachineRequirement(commandLineExecutor);
+        dockerMachineRequirement.check(createContext("testing"));
+    }
+
+    @Test(expected = UnsatisfiedRequirementException.class)
+    public void testDockerMachineRequirementCheckNoMachineFound() throws Exception {
+        when(commandLineExecutor.execCommandAsArray(anyVararg())).thenReturn(Arrays.asList(new String[] {"foo", "bar"}));
+
+        DockerMachineRequirement dockerMachineRequirement = new DockerMachineRequirement(commandLineExecutor);
+        dockerMachineRequirement.check(createContext(""));
+    }
+
+    @Test(expected = UnsatisfiedRequirementException.class)
+    public void testDockerMachineRequirementCheckNoMatchingNameNotMatched() throws Exception {
+        when(commandLineExecutor.execCommandAsArray(anyVararg())).thenReturn(Arrays.asList(new String[] {"my-docker-machine"}));
+
+        DockerMachineRequirement dockerMachineRequirement = new DockerMachineRequirement(commandLineExecutor);
+        dockerMachineRequirement.check(createContext("testing"));
+    }
+
+    @Test
+    public void testDockerMachineRequirementCheckNoMatchingNameMatched() throws Exception {
+        when(commandLineExecutor.execCommandAsArray(anyVararg())).thenReturn(Arrays.asList(new String[] {"testing"}));
+
+        DockerMachineRequirement dockerMachineRequirement = new DockerMachineRequirement(commandLineExecutor);
+        dockerMachineRequirement.check(createContext("testing"));
+    }
+
+    private RequiresDockerMachine createContext(String machineName) {
+        return new RequiresDockerMachine() {
+            @Override
+            public Class<? extends Annotation> annotationType() {
+                return RequiresDockerMachine.class;
+            }
+
+            @Override
+            public String name() {
+                return machineName;
+            }
+        };
+    }
+}

--- a/docker/docker/src/test/java/org/arquillian/cube/docker/impl/requirement/DockerRequirementTest.java
+++ b/docker/docker/src/test/java/org/arquillian/cube/docker/impl/requirement/DockerRequirementTest.java
@@ -1,0 +1,140 @@
+package org.arquillian.cube.docker.impl.requirement;
+
+import java.io.IOException;
+import java.io.OutputStreamWriter;
+import java.io.PrintWriter;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import org.arquillian.cube.docker.impl.client.CubeDockerConfiguration;
+import org.arquillian.cube.docker.impl.client.CubeDockerConfigurationResolver;
+import org.arquillian.cube.docker.impl.util.CommandLineExecutor;
+import org.arquillian.cube.spi.requirement.UnsatisfiedRequirementException;
+import org.arquillian.spacelift.execution.ExecutionException;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.runners.MockitoJUnitRunner;
+import static org.mockito.Matchers.anyMap;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class DockerRequirementTest {
+
+    @Mock
+    CubeDockerConfigurationResolver configResolver;
+
+    @Mock
+    CommandLineExecutor commandLineExecutor;
+
+    @Test(expected = UnsatisfiedRequirementException.class)
+    public void testDockerRequirementCheckWhenExecutionExceptionThrown() throws UnsatisfiedRequirementException {
+        when(configResolver.resolve(anyMap())).thenThrow(ExecutionException.class);
+
+        DockerRequirement requirement = new DockerRequirement(commandLineExecutor, configResolver);
+        requirement.check(null);
+    }
+
+    @Test(expected = UnsatisfiedRequirementException.class)
+    public void testDockerRequirementCheckWhenDockerURLIsNull() throws UnsatisfiedRequirementException {
+        Map<String, String> configMap = new HashMap<>();
+        configMap.put(CubeDockerConfiguration.DOCKER_URI, null);
+
+        when(configResolver.resolve(anyMap())).thenReturn(configMap);
+
+        DockerRequirement requirement = new DockerRequirement(commandLineExecutor, configResolver);
+        requirement.check(null);
+    }
+
+    @Test(expected = UnsatisfiedRequirementException.class)
+    public void testDockerRequirementCheckWhenDockerURLIsEmpty() throws UnsatisfiedRequirementException {
+        Map<String, String> configMap = new HashMap<>();
+        configMap.put(CubeDockerConfiguration.DOCKER_URI, "");
+
+        when(configResolver.resolve(anyMap())).thenReturn(configMap);
+
+        DockerRequirement requirement = new DockerRequirement(commandLineExecutor, configResolver);
+        requirement.check(null);
+    }
+
+    @Test(expected = UnsatisfiedRequirementException.class)
+    public void testDockerRequirementCheckWhenDockerURLIsInvalid() throws UnsatisfiedRequirementException {
+        Map<String, String> configMap = new HashMap<>();
+        configMap.put(CubeDockerConfiguration.DOCKER_URI, "tcp://nonexistanthostname:9999");
+
+        when(configResolver.resolve(anyMap())).thenReturn(configMap);
+
+        DockerRequirement requirement = new DockerRequirement(commandLineExecutor, configResolver);
+        requirement.check(null);
+    }
+
+    @Test
+    public void testDockerRequirementCheckDockerURLIsValid() throws Exception {
+        FakeDockerServer server = new FakeDockerServer();
+
+        Map<String, String> configMap = new HashMap<>();
+        configMap.put(CubeDockerConfiguration.DOCKER_URI, server.getConnectionString());
+
+        when(configResolver.resolve(anyMap())).thenReturn(configMap);
+
+        ExecutorService executorService = Executors.newSingleThreadExecutor();
+        executorService.submit(server);
+
+        try {
+            DockerRequirement requirement = new DockerRequirement(commandLineExecutor, configResolver);
+            requirement.check(null);
+        } finally {
+            executorService.shutdownNow();
+        }
+    }
+
+    private class FakeDockerServer implements Runnable {
+
+        private final ServerSocket serverSocket;
+
+        private FakeDockerServer() throws IOException {
+            serverSocket = new ServerSocket(0);
+        }
+
+        public String getConnectionString() {
+            return "tcp://" + serverSocket.getInetAddress().getHostName() + ":" + serverSocket.getLocalPort();
+        }
+
+        @Override
+        public void run() {
+            PrintWriter writer = null;
+            Socket socket = null;
+            try {
+                socket = serverSocket.accept();
+                writer = new PrintWriter(new OutputStreamWriter(socket.getOutputStream()));
+
+                String versionJSON = "{\"Client\":{\"Version\":\"0.0.0\",\"ApiVersion\":\"0.00\"}}";
+                writer.println("HTTP/1.1 200 OK");
+                writer.println("Content-Type: application/json");
+                writer.println("Content-Length: " + versionJSON.length());
+                writer.println();
+                writer.println(versionJSON);
+            } catch (IOException e) {
+                writer.println("HTTP/1.1 500");
+                writer.println();
+            } finally {
+                if (writer != null) {
+                    writer.close();
+                }
+
+                if (socket != null) {
+                    try {
+                        socket.close();
+                    } catch (IOException e) {
+                        return;
+                    }
+                }
+            }
+            return;
+        }
+    }
+}

